uber/models.py

```diff
@@ -1791,6 +1799,60 @@ def html(self):
             return SafeString(self.body.replace('\n', '<br/>'))
 
 
+class PromoCode(MagModel):
+    expiration_date = Column(UTCDateTime, default=c.ESCHATON)
+    price = Column(Integer, default=0)
+    code = Column(UnicodeText)
+    old_code = ""
+    # Uses should be -1 to make expired
+    uses = Column(Integer, nullable=True, default=None)
+    used_by = relationship(Attendee)
+
+    def generate_code(self, count):
+        code = []
+        for x in range(count):
+            code.append(c.PROMO_CODE_WORDS[random.choice(string.ascii_letters + string.digits)])
+        return "_".join(code)
+
+    @presave_adjustment
+    def _usage_count(self):
+        if self.uses == 0:
+            self.uses = None
+
+    @presave_adjustment
+    def _check_code(self):
+        if self.old_code == "":
```

Comment:
OK, it took a while but I think I finally get the point of the `old_code` thing. You want to check for duplicates in the system every time the code is saved, but only if the actual code has changed.

We actually have a function called `orig_value_of` and it does exactly what you want here. Just do this check:
`if self.code != self.orig_value_of('code')` and you can get rid of the `old_code` stuff you've put in.